Add media download functionality with :download command

Yash-007688 opened the pull request.

endernoke commented on source/utils/chat-commands.ts:
```diff
@@ -248,6 +248,46 @@ export const chatCommands: Record<string, ChatCommand> = {
+				return 'Usage: :select to enter selection mode first.';
+			}
+
+			const message = chatState.messages[chatState.selectedMessageIndex];
```
`message` may be undefined here, see the unsend command implementation for a working type check

endernoke commented on source/client.ts:
```diff
@@ -919,6 +919,86 @@ export class InstagramClient extends EventEmitter {
+				}
+				
+				try {
+					await streamPipeline(response, fsPromises.createWriteStream(downloadPath));
```
There's a build error here:

```
source/client.ts:947:48 - error TS2339: Property 'createWriteStream' does not exist on type '{ default: typeof import("node:fs/promises"); access(path: PathLike, mode?: number | undefined): Promise<void>; copyFile(src: PathLike, dest: PathLike, mode?: number | undefined): Promise<...>; ... 30 more ...; constants: typeof constants; }'.

947      await streamPipeline(response, fsPromises.createWriteStream(downloadPath));
                                                   ~~~~~~~~~~~~~~~~~


```

Yash-007688 commented:
Hi James,

Thank you for the review and for the LGTM 👍 — I really appreciate it.

I’ve addressed all the requested changes:

   -

   Added proper type-safe checks for the selected message (following the
   unsend command pattern)
   -

   Moved all dynamic imports to static top-level imports
   -

   Fixed the TypeScript build issue by using fs.createWriteStream with
   stream.pipeline
   -

   Verified that all tests pass locally using npm test, and formatting
   checks are clean

Please let me know if there’s anything else you’d like me to adjust.
Thanks again for taking the time to review this!

Best regards,
Yash
(GitHub: Yash-007688)

On Tue, Dec 30, 2025 at 1:32 PM James Zheng ***@***.***>
wrote:

> ***@***.**** requested changes on this pull request.
>
> @Yash-007688 <https://github.com/Yash-007688> Overall LGTM 👍 Thanks for
> working on this.
>
> I've left some comments and also there're a few type errors. Please make
> sure you verify that all tests pass and fix all issues that arise:
>
> npm test
>
> ------------------------------
>
> In source/utils/chat-commands.ts
> <https://github.com/supreme-gg-gg/instagram-cli/pull/218#discussion_r2652450021>
> :
>
> > @@ -248,6 +248,46 @@ export const chatCommands: Record<string, ChatCommand> = {
>  			return;
>  		},
>  	},
> +	download: {
> +		description: 'Download media from the selected message. Usage: :download [path]',
> +		async handler(arguments_, {client, chatState, setChatState}) {
> +			if (chatState.selectedMessageIndex === undefined) {
> +				return 'Usage: :select to enter selection mode first.';
> +			}
> +
> +			const message = chatState.messages[chatState.selectedMessageIndex];
>
> message may be undefined here, see the unsend command implementation for
> a working type check
> ------------------------------
>
> In source/utils/chat-commands.ts
> <https://github.com/supreme-gg-gg/instagram-cli/pull/218#discussion_r2652450961>
> :
>
> > +
> +			const message = chatState.messages[chatState.selectedMessageIndex];
> +			if (message.itemType !== 'media' || !message.media) {
> +				return 'Selected message does not contain media.';
> +			}
> +
> +			if (!chatState.currentThread) {
> +				return;
> +			}
> +
> +			// Determine default download path if not provided
> +			const downloadPath = arguments_[0] || `./downloads/media_${message.id}`;
> +			
> +			// Create downloads directory if it doesn't exist
> +			const fs = await import('node:fs');
> +			const path = await import('node:path');
>
> Please move these to static imports at the top of the file. Also applies
> to client.ts
> ------------------------------
>
> In source/client.ts
> <https://github.com/supreme-gg-gg/instagram-cli/pull/218#discussion_r2652455757>
> :
>
> > +			const fsPromises = await import('node:fs/promises');
> +			const url = await import('node:url');
> +			
> +			const streamPipeline = promisify(pipeline);
> +			const parsedUrl = new url.URL(mediaUrl);
> +			const client = parsedUrl.protocol === 'https:' ? https : http;
> +			
> +			return new Promise((resolve, reject) => {
> +				const request = client.get(mediaUrl, async (response) => {
> +					if (response.statusCode !== 200) {
> +						reject(new Error(`Failed to download media: ${response.statusCode} ${response.statusMessage}`));
> +						return;
> +				}
> +				
> +				try {
> +					await streamPipeline(response, fsPromises.createWriteStream(downloadPath));
>
> There's a build error here:
>
> source/client.ts:947:48 - error TS2339: Property 'createWriteStream' does not exist on type '{ default: typeof import("node:fs/promises"); access(path: PathLike, mode?: number | undefined): Promise<void>; copyFile(src: PathLike, dest: PathLike, mode?: number | undefined): Promise<...>; ... 30 more ...; constants: typeof constants; }'.
>
> 947      await streamPipeline(response, fsPromises.createWriteStream(downloadPath));
>                                                    ~~~~~~~~~~~~~~~~~
>
>
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/supreme-gg-gg/instagram-cli/pull/218#pullrequestreview-3617507925>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/BQMBPS26MWPK2ZOS7MTNXQD4EIWRHAVCNFSM6AAAAACQJICASKVHI2DSMVQWIX3LMV43YUDVNRWFEZLROVSXG5CSMV3GSZLXHMZTMMJXGUYDOOJSGU>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

supreme-gg-gg reviewed (changes requested):
Some minor picks, the overall functionality is working now

supreme-gg-gg commented on tsconfig.json:
```diff
@@ -9,9 +9,15 @@
+		"strictNullChecks": true
 	},
-	"include": ["./source"],
+	"include": ["./source/**/*", "./xo.config.ts", "./tests/**/*"],
```
@Yash-007688 can you remove changes to `tsconfig.json`? (perhaps except the `exclude` field which is reasonable to keep)

supreme-gg-gg commented:
@Yash-007688 why is `tsconfig.json` a binary file? can you just restore this file to the version in main? also, you might need to resolve conflicts in `chat-commands.ts` before I can merge this! You can do `git fetch upstream main && git rebase upstream/main` or something equivalent
